src/autopas/containers/octree/Octree.h

```diff
@@ -0,0 +1,240 @@
+ *
+ * @tparam Particle
+ */
+template <class Particle>
+class Octree : public CellBasedParticleContainer<OctreeNodeWrapper<Particle>>,
+               public internal::CellBorderAndFlagManager {
+ public:
+  /**
+   * The particle cell used in this CellBasedParticleContainer
+   */
+  using ParticleCell = OctreeNodeWrapper<Particle>;
+
+  /**
+   * The particle type used in this container.
+   */
+  using ParticleType = typename ParticleCell::ParticleType;
+
+  /**
+   * This particle container contains two cells. Both cells are octrees.
+   * - this->_cells[CellTypes::OWNED] yields the octree that contains all owned particles
+   * - this->_cells[CellTypes::HALO] yields the octree that contains all halo particles
+   */
+  enum CellTypes { OWNED = 0, HALO = 1 };
+
+  /**
+   * Construct a new octree with two sub-octrees: One for the owned particles and one for the halo particles.
+   * @param boxMin The minimum coordinate of the enclosing box
+   * @param boxMax The maximum coordinate of the enclosing box
+   * @param cutoff The cutoff radius
+   * @param skin The skin radius
+   * @param cellSizeFactor The cell size factor
+   */
+  Octree(std::array<double, 3> boxMin, std::array<double, 3> boxMax, const double cutoff, const double skin,
+         const double cellSizeFactor)
+      : CellBasedParticleContainer<ParticleCell>(boxMin, boxMax, cutoff, skin) {
+    // TODO(johannes): Obtain this from a configuration, reported in https://github.com/AutoPas/AutoPas/issues/624
+    int unsigned treeSplitThreshold = 16;
+
+    double interactionLength = this->getInteractionLength();
+    this->_cells.push_back(
+        OctreeNodeWrapper<Particle>(boxMin, boxMax, treeSplitThreshold, interactionLength, cellSizeFactor));
+
+    // Extend the halo region with cutoff + skin in all dimensions
+    auto haloBoxMin = utils::ArrayMath::subScalar(boxMin, this->getInteractionLength());
+    auto haloBoxMax = utils::ArrayMath::addScalar(boxMax, this->getInteractionLength());
+    this->_cells.push_back(
+        OctreeNodeWrapper<Particle>(haloBoxMin, haloBoxMax, treeSplitThreshold, interactionLength, cellSizeFactor));
+  }
+
+  [[nodiscard]] std::vector<ParticleType> updateContainer() override {
+    // This is a very primitive and inefficient way to recreate the container:
+    // 1. Copy all particles out of the container
+    // 2. Clear the container
+    // 3. Insert the particles back into the container
+
+    // leaving: all outside boxMin/Max
+
+    // TODO(johannes): Make this less indirect. (Find a better way to iterate all particles inside the octree to change
+    //  this function back to a function that actually copies all particles out of the octree.)
+    //  The problem is captured by https://github.com/AutoPas/AutoPas/issues/622
+    std::vector<Particle *> particleRefs;
+    this->_cells[CellTypes::OWNED].appendAllParticles(particleRefs);
+    std::vector<Particle> particles;
+    auto result = std::vector<ParticleType>();
```

Comment:
```suggestion
    std::vector<Particle> particles{};
    particles.reserve(particleRefs.size());
    std::vector<Particle> result{};
```
maybe also rename `result` to `invalidParticles` or similar?
Also: Most particles will end up in the `particles` vector, hence I think it would make sense to do some reserving.

Reply:
Adopted changes in 981febf.